### PR TITLE
Element\Select selecting incorrect values when using floats (issue #52)

### LINKED ISSUE #52: When calling setValue Element\Select, selected incorrect values
When calling $select->setValue(array('1.1')), selected values French and Japanese,
but Japanese is incorrect value

``` php
use Zend\Form\Element;
use Zend\Form\Form;

$select = new Element\Select('language');
$select->setLabel('Which is your mother tongue?');
$select->setValueOptions(array(
    '1.1' => 'French',
    '1.2' => 'English',
    '1.10' => 'Japanese',
    '1.20' => 'Chinese',
));

$form = new Form('language');
$form->add($select);
```

---

Originally posted by @googlle at https://github.com/zendframework/zend-form/issues/18

### PR DESCRIPTION
**Pull request by driehle**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This MR fixes #52 (and zendframework/zend-form#18) by adding the test case provided in #52 and an implementation to let that test pass.

I would consider this a BC break since as people who might have relied on the currently implemented weak comparison in v2.x might be affected by this change. However, that will probably be an edge case.

### CODE REVIEW

**Comment by froschdesign** [comment]
@driehle 
It looks like there is too much in the branch. 😉

**Comment by driehle** [comment]
> It looks like there is too much in the branch. 😉

I purposely based it on #113 since that one is already in review. Once that is merged only one commit will remain here. :-)

**Comment by Slamdunk** [comment]
@driehle please rebase this PR

**Comment by driehle** [comment]
done

File: src/View/Helper/FormSelect.php [hunk 1]
```diff
@@ -205,7 +204,7 @@ public function renderOptions(array $options, array $selectedOptions = [])
                 $disabled = $optionSpec['disabled'];
             }
 
-            if (ArrayUtils::inArray($value, $selectedOptions)) {
+            if ($this->inArrayWithStringConversion($value, $selectedOptions)) {
```
Comment (Slamdunk) [inline src/View/Helper/FormSelect.php]: There is no need for a separate method, just use the `strict` variant: `ArrayUtils::inArray($value, $selectedOptions, true)`

**Review by Slamdunk — CHANGES_REQUESTED** [review]

File: src/View/Helper/FormSelect.php (hunk 1 above)
Comment (driehle) [reply src/View/Helper/FormSelect.php]: Unfortunatly, that won't work. Strict comparison failes because the unit tests assume that objects are identical when their string representation equals, but enabling strict comparison requires objects to be identical. Hence, if I replace the line as you suggested, the following unit tests fail:

```
1) LaminasTest\Form\View\Helper\FormSelectTest::testCanMarkOptionsAsSelectedWhenEmptyOptionOrZeroValueSelected
Failed asserting that '<select name="foo"><option value="">empty</option>\n
<option value="0">label0</option>\n
<option value="1">label1</option></select>' contains "<option value="0" selected="selected">label0</option>".

/home/driehle/PhpstormProjects/laminas-form/test/View/Helper/FormSelectTest.php:390

2) LaminasTest\Form\View\Helper\FormSelectTest::testRenderElementWithObjectIdentifiers
Failed asserting that '<select name="foo"><option value="42">This is the first label</option>\n
<option value="43">This is the second label</option></select>' matches PCRE pattern "#option .*?value="42" selected="selected"#".

/home/driehle/PhpstormProjects/laminas-form/test/View/Helper/FormSelectTest.php:441
```

To circumstance this, I have added the new method, which first converts objects to string and then performs a strict comparison. Such functionality is not provided in stdlib, I think.

Comment (Slamdunk) [reply src/View/Helper/FormSelect.php]: You are right, thank you
